### PR TITLE
Flags backend

### PR DESCRIPTION
**Pull request by g-ortuno**
Added ability to change the `Heart Rate Measurement Value` and `Energy Expended` fields of the `Heart Rate Measurement` characteristic from the UI and notify connected devices of the change. Please review @jyasskin @scheib

### CODE REVIEW

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java [hunk 1]
```diff
@@ -61,18 +157,75 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
+     * we can use setValue(value, format, offset);
+     * Flags (8bit) + Heart Rate Measurement Value (uint8) + Energy Expended (uint16) = 4 bytes
+     *
+     * Flags = 1 << 3:
+     * Heart Rate Format (0) -> UINT8
+     * Sensor Contact Status (00) -> Not Supported
+     * Energy Expended (1) -> Field Present
+     * RR-Interval (0) -> Field not pressent
+     */
+    mHeartRateMeasurementCharacteristic.setValue(new byte[]{1 << 3, 0, 0, 0});
```
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Does `0b01000` work here?

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java [hunk 2]
```diff
@@ -28,10 +53,68 @@
...
+          if (newEnergyExpended <= MAX_ENERGY_EXPENDED) {
+            mHeartRateMeasurementCharacteristic.setValue(newEnergyExpended,
+                BluetoothGattCharacteristic.FORMAT_UINT16,
+                /* offset */ 2);
+          } else {
+            Toast.makeText(getActivity(), R.string.energyExpendedTooHigh,
+                Toast.LENGTH_SHORT).show();
+          }
+        } else {
+          Toast.makeText(getActivity(), R.string.energyExpendedInvalid,
```
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: I would put these closer to the condition that triggers them. So:
```java
if (newEnergyExpendedString.isEmpty() ||
    !TextUtils.isDigitsOnly(newEnergyExpendedString) {
  Toast.makeText(getActivity(), R.string.energyExpendedInvalid,
      Toast.LENGTH_SHORT).show();
  return false;
}
if (newEnergyExpended > MAX_ENERGY_EXPENDED) {
  Toast.makeText(getActivity(), R.string.energyExpendedTooHigh,
      Toast.LENGTH_SHORT).show();
  return false;
}
```

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java [hunk 3]
```diff
@@ -28,10 +53,68 @@
...
+  private final OnEditorActionListener mOnEditorActionListenerEnergyExpended = new OnEditorActionListener() {
+    @Override
+    public boolean onEditorAction(TextView textView, int actionId, KeyEvent event) {
+      if (actionId == EditorInfo.IME_ACTION_DONE) {
+        String newEnergyExpendedString = textView.getText().toString();
+        // Need to check if the string is empty since isDigitsOnly returns
+        // true for empty strings.
+        if (!newEnergyExpendedString.isEmpty()
+            && TextUtils.isDigitsOnly(newEnergyExpendedString)) {
+          int newEnergyExpended = Integer.parseInt(newEnergyExpendedString);
```
Comment (jyasskin) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Do you want to catch the `NumberFormatException` caused by a too-big integer?

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 1 above)
Comment (scheib) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Comment explains the array easily, but the flags aren't quite clear without reading spec, here's a small change with indents and the unused bits also that would have made it clear to me quickly:

```
+    /* Set the org.bluetooth.characteristic.heart_rate_measurement
+     * characteristic to a byte array of size 4 so
+     * we can use setValue(value, format, offset);
+     *
+     * Flags (8bit) + Heart Rate Measurement Value (uint8) + Energy Expended (uint16) = 4 bytes
+     *
+     * Flags = 1 << 3:
+     *   Heart Rate Format (0) -> UINT8
+     *   Sensor Contact Status (00) -> Not Supported
+     *   Energy Expended (1) -> Field Present
+     *   RR-Interval (0) -> Field not pressent
+     *   Unused (000)
```

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java [hunk 4]
```diff
@@ -61,18 +157,75 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
+     * Flags (8bit) + Heart Rate Measurement Value (uint8) + Energy Expended (uint16) = 4 bytes
+     *
+     * Flags = 1 << 3:
+     * Heart Rate Format (0) -> UINT8
+     * Sensor Contact Status (00) -> Not Supported
+     * Energy Expended (1) -> Field Present
+     * RR-Interval (0) -> Field not pressent
+     */
+    mHeartRateMeasurementCharacteristic.setValue(new byte[]{1 << 3, 0, 0, 0});
+    // Characteristic Value: [8, 0, 0, 0]
```
Comment (scheib) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: I'm having a hard time following the // Characteristic Value: comments. Where are the 8 and 60 values coming from?

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 1 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Yes it does :)
Done.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 2 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Rewrote this part. Please review again.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 3 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: This is handled by `isValidCharacteristicValue()` now.

File: app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java (hunk 4 above)
Comment (g-ortuno) [inline app/src/main/java/io/github/webbluetoothcg/bletestperipheral/HeartRateServiceFragment.java]: Done.

**Comment by g-ortuno** [comment]
Ready for review again!

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -35,4 +35,6 @@
     <string name="heartRateServiceName">Heart Rate Service</string>
     <string name="label_heartRateMeasurementValue">Heart Rate</string>
     <string name="label_energyExpended">Energy Expended</string>
+    <string name="heartRateMeasurementValueInvalid">Please enter a number between 0 and 255</string>
+    <string name="energyExpendedInvalid">Please enter a number between 0 and 65,535</string>
```
Comment (jyasskin) [inline app/src/main/res/values/strings.xml]: I'd probably skip the comma in `65535` since the user isn't allowed to actually type it.

**Comment by jyasskin** [comment]
LGTM

**Pull request merged**